### PR TITLE
#2979 - Email notification - PD/PPD Student reminder email 8 weeks before end date

### PR DESCRIPTION
**Pull request by bidyashish**
## Scheduler setup
- [x] DB Migration for the new Scheduler
- [x] Scheduler cron set for `08:00 AM UTC  Every Day`
image

- [x] New scheduler Class and related objects with TODO implementation and basic log.

Migrations Revert Demo

image

Unchecked AC for TODO
- [X] Create a scheduler to verify the applications once a day. The scheduler can be generic to support similar validations in the future.
  - Suggested name: `student-application-notifications`.
  - Pick a time after `archive-applications` with 30min to 1h difference.
- [X] Add the new scheduler for the wiki. https://github.com/bcgov/SIMS/wiki/Schedulers
- [ ] Select all student applications where
  - Offerings end date within 8 weeks.
  - Applications with current disbursement pending.
  - Application not archived.
  - Remove assessments where a notification was already sent.
  - With the PD/PPD mismatch using the same logic from [validate-disbursement-base.ts](https://github.com/bcgov/SIMS/blob/main/sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/validate-disbursement-base.ts#L44).
- [ ] Save the assessment ID as metadata and ensure the email will be sent once.
- [ ] E2E Tests
  - [ ] Ensure the email will be sent once for the same assessment.
  - [ ] Ensure the email will be sent again for different assessments for the same application.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/_tests_/student-application-notifications.scheduler.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+describe("StudentApplicationNotificationsScheduler (e2e)", () => {
```
Comment (guru-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/_tests_/student-application-notifications.scheduler.e2e-spec.ts:1]: Please remove this dummy test file as, this does not have any coverage for the files in the PR

File: sources/packages/backend/apps/db-migrations/src/migrations/1727119765682-StudentApplicationNotifications.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      getSQLFileData(
+        "Add-student-application-notifications-queue.sql",
+        "Queue",
+      ),
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
```
Comment (guru-aot) [inline sources/packages/backend/apps/db-migrations/src/migrations/1727119765682-StudentApplicationNotifications.ts:16]: Please have the correct file name in the rollback script.

**Comment by guru-aot** [comment]
Please always add proof of rollback during the PR review so that it would help the PR reviewers to have an idea that the migration and its revert is successful.

image
 In this case the above revert is failing.

**Comment by guru-aot** [comment]
Please check the sonar cloud issues too.

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-student-application-notifications-queue.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
+INSERT INTO
```
Comment (guru-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-student-application-notifications-queue.sql:1]: Formatting is not right, please check if all the plugins for formatter is in place.
https://github.com/bcgov/SIMS/wiki/Development---Standards,-Decisions,-and-Troubleshooting#plugins-to-be-installed

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts [hunk 4]
```diff
@@ -153,4 +153,16 @@ export class ApplicationService {
         .getMany()
     );
   }
+
+  async getEligibleApplicationsForNotification(): Promise<Application[]> {
```
Comment (guru-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:157]: If the PR is related to only scheduler creation, please remove these files from the PR.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import { NotificationService } from "@sims/services/notifications";
+
+@Processor(QueueNames.StudentApplicationNotifications)
+export class StudentApplicationNotificationsScheduler extends BaseScheduler<void> {
+  constructor(
+    @InjectQueue(QueueNames.StudentApplicationNotifications)
+    schedulerQueue: Queue<void>,
+    queueService: QueueService,
+    private readonly applicationService: ApplicationService,
+    private readonly notificationService: NotificationService,
```
Comment (guru-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:25]: I believe this file can also be created as part of the upcoming PR and can be removed from this PR. Please remove it. Else unwanted declared not used variables will be in the main branch, once merged.

**Review by guru-aot — CHANGES_REQUESTED** [review]
Had a look on the PR, i believe overall it would be better to remove the unwanted files which are not complete and checking the migration revert would be better with proof. Nice work @bidyashish

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      processSummary.info(` ${eligibleApplications}`);
+
+      return getSuccessMessageWithAttentionCheck(
+        ["Process finalized with success."],
+        processSummary,
+      );
+    } catch (error: unknown) {
+      const errorMessage = "Unexpected error while executing the job.";
+      processSummary.error(errorMessage, error);
+      return [errorMessage];
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:52]: Please throw the error here to allow the job to retry.
```ts
throw new Error(errorMessage, { cause: error });
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    try {
+      this.logger.log(
+        `Processing student application notifications job. Job id: ${job.id} and Job name: ${job.name}.`,
+      );
+
+      const eligibleApplications =
+        await this.applicationService.getEligibleApplicationsForNotification();
+      // TODO: Get applications that have a disability status mismatch  and check PDPPD status
+
+      processSummary.info(` ${eligibleApplications}`);
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:43]: Please remove this code nd leave just a TODO.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  @Process()
+  async StudentApplicationNotifications(job: Job<void>): Promise<string[]> {
+    const processSummary = new ProcessSummary();
+
+    try {
+      this.logger.log(
+        `Processing student application notifications job. Job id: ${job.id} and Job name: ${job.name}.`,
+      );
+
+      const eligibleApplications =
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:39]: The method `this.applicationService.getEligibleApplicationsForNotification()` will likely be called from another service responsible for processing the scheduler logic.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    schedulerQueue: Queue<void>,
+    queueService: QueueService,
+    private readonly applicationService: ApplicationService,
+    private readonly notificationService: NotificationService,
+  ) {
+    super(schedulerQueue, queueService);
+  }
+
+  @Process()
+  async StudentApplicationNotifications(job: Job<void>): Promise<string[]> {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:31]: Please rename it to `studentApplicationNotifications`.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    @InjectQueue(QueueNames.StudentApplicationNotifications)
+    schedulerQueue: Queue<void>,
+    queueService: QueueService,
+    private readonly applicationService: ApplicationService,
+    private readonly notificationService: NotificationService,
+  ) {
+    super(schedulerQueue, queueService);
+  }
+
+  @Process()
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:30]: Please add some comments.

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:157]: If the file is present in the PR it must be good enough to be merged to main. No temporary or in-progress code should be present unless it is a partial implementation of the final approach, wich does not seem to be the case here.

**Review by andrewsignori-aot — COMMENTED** [review]
Please take a look at the comments.
Please review the below PR decisions:
https://github.com/bcgov/SIMS/wiki/Development---Standards,-Decisions,-and-Troubleshooting#pr-review
It contains a specific topic about
https://github.com/bcgov/SIMS/wiki/Development---Standards,-Decisions,-and-Troubleshooting#migration-rollback-scripts

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-student-application-notifications-queue.sql (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-student-application-notifications-queue.sql:1]: The mentioned extension is the [SQL Formatter](https://marketplace.visualstudio.com/items?itemName=adpyke.vscode-sql-formatter) mentioned in the link shared by @guru-aot.
The format change is minor for this SQL, but not having it installed will be a problem for more complex queries. Please ensure that it is installed and formatting the code while saving.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    private readonly applicationService: ApplicationService,
+    private readonly notificationService: NotificationService,
+  ) {
+    super(schedulerQueue, queueService);
+  }
+
+  /**
+   * Process Student for Email notification - PD/PPD Student reminder email 8 weeks before end date
+   */
+
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:33]: Please remove the blank line.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    schedulerQueue: Queue<void>,
+    queueService: QueueService,
+    private readonly applicationService: ApplicationService,
+    private readonly notificationService: NotificationService,
+  ) {
+    super(schedulerQueue, queueService);
+  }
+
+  /**
+   * Process Student for Email notification - PD/PPD Student reminder email 8 weeks before end date
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts:31]: Please finish the sentences with a period.

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for making the changes and please take a look in the latest comments.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/student-application-notifications/student-application-notifications.scheduler.ts]: I will wait to see the next PR to have a better context.
Please keep in mind that this PR is to send one notification based on some criteria and that we are expecting other application-based notifications to be created in the future.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, look good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @bidyashish

**Pull request merged**